### ISSUE TITLE
diffHashes method in Table not working properly

### ISSUE THREAD

**Issue by torbjornvatn** [cucumber-attic/cuke4duke#18]
**UPDATED**

I tried to use the diffHashes method of the Table class like this: http://gist.github.com/184345

When I run the steps I get this error:

```
 [INFO]     When I add a table               
 [INFO]       | a | b | b2 | a1 |
 [INFO]       | 1 | 2 |    |    |
 [INFO]       Tables were not identical (RuntimeError)
 [INFO]       features/demo.feature:13:in `When I add a table'
```

The last two columns are gray (surplus)

**Comment by aslakhellesoy** [comment]
What colours do you get? One row grey (surplus) and one yellow (missing)? In that order or the opposite? It could be a cucumber bug, but it could also be a cuke4duke bug.

Wanna investigate?

**Comment by torbjornvatn** [comment]
The second | 1 | 2 | is gray (surplus).
When I print the List<Map<String, String>> I made it looks like: [a1b2, {b=2, a=1}]
When I print table.hashes it looks like: a1b2

Could it be a JRuby RubyArray vs. Java ArrayList problem?

Will try to investigate, at least on the cuke4duke side. The ruby code in cucumber is a bit hardcore for me as a ruby novice

**Comment by aslakhellesoy** [comment]
Cool. The elimination technique works well here. If you can confirm it's _not_ a cuke4duke bug I'll investigate on the cucumber/ruby side. Otherwise you'll fix the cuke4duke bug ;-)

**Comment by darrinholst** [comment]
This is the first time I've looked at the diff stuff so maybe I'm missing something, but isn't that what is supposed to happen?

You initialized the list with what was already in the table and then added a new map giving you a surplus row

```
  # An exception is raised if there are missing rows or columns, or
  # surplus rows. An error is <em>not</em> raised for surplus columns.
  # Whether to raise or not raise can be changed by setting values in
  # +options+ to true or false:
```

It does look like the ruby hash prints differently then the java map, but I think it all works itself out by the time it gets to ruby land.

**Comment by torbjornvatn** [comment]
Sorry about that,it was a copy/paste error in the last gist. Corrected here: http://gist.github.com/184345
The new error I get when this code is run looks like this:

```
[INFO]     When I add a table               
[INFO]       | a | b | b2 | a1 |
[INFO]       | 1 | 2 |    |    |
[INFO]       Tables were not identical (RuntimeError)
[INFO]       features/demo.feature:13:in `When I add a table'
```

The two new columns are gray (surplus)

**Comment by darrinholst** [comment]
It appears to be the following method in cucumber where it converts the input to a Table

```
  def ensure_table(table_or_array) #:nodoc:
    puts table_or_array.class    #=> Java::JavaUtil::ArrayList
    puts table_or_array[0]       #=> {b=2, a=1} 
    puts table_or_array[0].class #=> Java::JavaUtil::HashMap
    return table_or_array if Table === table_or_array
    table_or_array = hashes_to_array(table_or_array) if Hash === table_or_array[0]
    table_or_array = enumerable_to_array(table_or_array) unless Array == table_or_array[0]
    Table.new(table_or_array)
  end
```

What is returned is 
  | b2 | a2 |

**Comment by aslakhellesoy** [comment]
Thanks Darrin. Looks like this method needs to be smarter in order to cater for Java collections. Alternatively, we could try to convert to Ruby types on the cuke4duke side before sending it down. Any suggestions?

**Comment by darrinholst** [comment]
Yeah, I'd hate to have to change cucumber for this. Could we open up === in java::util::HashMap and handle it there? I couldn't find an easy way to convert java maps to ruby hashes. Would you just have to manually iterate over it?

**Comment by aslakhellesoy** [comment]
We _could_ open up java::util::HashMap#===, but it's hacky. I'd prefer looking from some pertinent methods with respond_to? that could be used to separate "hashish", "arrayish" and "tableish" :-)

**Comment by darrinholst** [comment]
Actually you'd have to open up Hash to handle HashMaps, but that screws up the === check when you're converting the options.

respond_to? was actually my first thought, but I can't find any common methods between Hash and HashMap. We could add a to_hash or alias contains_key? to has_key? in HashMap and check for those in cucumber.

**Comment by aslakhellesoy** [comment]
Doesn' java.util.Map classes respond to the same methods as Ruby Hash? At least you can _use_ a java.util.Map as if it were a Hash.

**Comment by darrinholst** [comment]
No, the jruby wrapper for a java Map just mixes in Enumerable, good article on that here http://www.engineyard.com/blog/2009/youve-got-java-in-my-ruby/ in the Decoration section.

**Comment by darrinholst** [comment]
So I think we can do something like the following in cucumber_ext.rb

  module java::util::Map
    alias_method :has_key?, :contains_key?
  end

**Comment by darrinholst** [comment]
here's what I tried http://gist.github.com/184755, table.rb changed in cucumber and cucumber_ext.rb and StuffSteps.java changed in cuke4duke

Now I'm getting 

When I add a table               # public void simple.StuffSteps.iAddATable(cuke4duke.Table)

| a | b |

| 1 | 2 |

| 1 | 2 |

Tables were not identical (RuntimeError)

features/demo.feature:13:in `When I add a table'

Where the 3rd row is yellow and when I print the converted table out now it's

 |     b |     a |

|     2 |     1 |

I tried figuring out the cucumber code where it's doing the diff and got lost, is it doing a text comparison to check for missing/surplus rows? The keys ('a', 'b') returned from the Java map come back in a different order than a ruby collection.

Any ideas? Or should we just go with converting java collections to ruby before calling diff?

**Comment by darrinholst** [comment]
I went with the java map to ruby conversion - check it out at http://github.com/darrinholst/cuke4duke/commit/36e1a3fa8b89fb913b78f29ed38dbd12e9b8b8ef

It worked on the project I was trying to get this working on, I can update History.txt if you want to go with it.

**Comment by aslakhellesoy** [comment]
diffHashes method in Table not working properly. Closed by b07d8e5ec90963ffa572597ac2612aed5ca7d5e4
